File: Classes/View Controllers/Tabs/Folders/FolderAlbumViewController.m
//
//  AlbumViewController.m
//  iSub
//
//  Created by Ben Baron on 2/28/10.
//  Copyright 2010 Ben Baron. All rights reserved.
//

#import "FolderAlbumViewController.h"
#import "UIViewController+PushViewControllerCustom.h"
#import "iSubAppDelegate.h"
#import "ViewObjectsSingleton.h"
#import "Defines.h"
#import "SavedSettings.h"
#import "MusicSingleton.h"
#import "DatabaseSingleton.h"
#import "EX2Kit.h"
#import "Swift.h"
#import <MediaPlayer/MediaPlayer.h>

@interface FolderAlbumViewController() <APILoaderDelegate>
@end

@implementation FolderAlbumViewController

#pragma mark Lifecycle

- (instancetype)initWithFolderArtist:(ISMSFolderArtist *)folderArtist orFolderAlbum:(ISMSFolderAlbum *)folderAlbum {
	if (folderArtist == nil && folderAlbum == nil) return nil;
	
    if (self = [super init]) {
		if (folderArtist != nil) {
			self.title = folderArtist.name;
			_folderId = folderArtist.folderId;
			_folderArtist = folderArtist;
		} else {
			self.title = folderAlbum.name;
			_folderId = folderAlbum.folderId;
			_folderAlbum = folderAlbum;
		}
        _dataModel = [[SubfolderDAO alloc] initWithParentFolderId:_folderId delegate:self];
		
        if (self.dataModel.hasLoaded) {
            [self.tableView reloadData];
            [self addHeaderAndIndex];
        } else {
            [viewObjectsS showAlbumLoadingScreen:self.view sender:self];
            [self.dataModel startLoad];
        }
	}
	
	return self;
}

- (void)viewDidLoad  {
    [super viewDidLoad];
    				
    // Add the pull to refresh view
    __weak FolderAlbumViewController *weakSelf = self;
    self.refreshControl = [[RefreshControl alloc] initWithHandler:^{
        [viewObjectsS showAlbumLoadingScreen:weakSelf.view sender:weakSelf];
        [weakSelf.dataModel startLoad];
    }];
    
    self.tableView.rowHeight = Defines.rowHeight;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
    self.tableView.separatorColor = UIColor.clearColor;
    [self.tableView registerClass:UniversalTableViewCell.class forCellReuseIdentifier:UniversalTableViewCell.reuseId];
}

- (void)reloadData {
    [self.tableView reloadData];
}

- (void)viewWillAppear:(BOOL)animated  {
	[super viewWillAppear:animated];
	
	if (musicS.showPlayerIcon) {
		self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage systemImageNamed:@"music.quarternote.3"] style:UIBarButtonItemStylePlain target:self action:@selector(nowPlayingAction:)];
	} else {
		self.navigationItem.rightBarButtonItem = nil;
	}
	
	[self.tableView reloadData];
		
	[NSNotificationCenter addObserverOnMainThread:self selector:@selector(reloadData) name:ISMSNotification_CurrentPlaylistIndexChanged];
	[NSNotificationCenter addObserverOnMainThread:self selector:@selector(reloadData) name:ISMSNotification_SongPlaybackStarted];
}

- (void)viewWillDisappear:(BOOL)animated {
	[super viewWillDisappear:animated];
	
	[self.dataModel cancelLoad];
	
	[NSNotificationCenter removeObserverOnMainThread:self name:ISMSNotification_CurrentPlaylistIndexChanged];
	[NSNotificationCenter removeObserverOnMainThread:self name:ISMSNotification_SongPlaybackStarted];
}

- (void)dealloc {
	[NSNotificationCenter removeObserverOnMainThread:self];
	self.dataModel.delegate = nil;
}

#pragma mark Loading

- (void)cancelLoad {
	[self.dataModel cancelLoad];
    [self.refreshControl endRefreshing];
	[viewObjectsS hideLoadingScreen];
}

// Autolayout solution described here: https://medium.com/@aunnnn/table-header-view-with-autolayout-13de4cfc4343
- (void)addHeaderAndIndex {
	if (self.dataModel.songCount == 0 && self.dataModel.folderCount == 0) {
		self.tableView.tableHeaderView = nil;
    } else {
        // Create the container view and constrain it to the table
        UIView *headerView = [[UIView alloc] init];
        headerView.translatesAutoresizingMaskIntoConstraints = NO;
        self.tableView.tableHeaderView = headerView;
        [NSLayoutConstraint activateConstraints:@[
            [headerView.centerXAnchor constraintEqualToAnchor:self.tableView.centerXAnchor],
            [headerView.widthAnchor constraintEqualToAnchor:self.tableView.widthAnchor],
            [headerView.topAnchor constraintEqualToAnchor:self.tableView.topAnchor]
        ]];
        
        // Create the play all and shuffle buttons and constrain to the container view
        __weak FolderAlbumViewController *weakSelf = self;
        PlayAllAndShuffleHeader *playAllAndShuffleHeader = [[PlayAllAndShuffleHeader alloc] initWithPlayAllHandler:^{
            [SongLoader playAllWithFolderId:weakSelf.folderId];
        } shuffleHandler:^{
            [SongLoader shuffleAllWithFolderId:weakSelf.folderId];
        }];
        [headerView addSubview:playAllAndShuffleHeader];
        [NSLayoutConstraint activateConstraints:@[
            [playAllAndShuffleHeader.leadingAnchor constraintEqualToAnchor:headerView.leadingAnchor],
            [playAllAndShuffleHeader.trailingAnchor constraintEqualToAnchor:headerView.trailingAnchor],
            [playAllAndShuffleHeader.bottomAnchor constraintEqualToAnchor:headerView.bottomAnchor]
        ]];
        
        if (self.dataModel.songCount > 0) {
            // Create the album header view and constrain to the container view
            AlbumTableViewHeader *albumHeader = [[AlbumTableViewHeader alloc] initWithFolderAlbum:self.folderAlbum tracks:self.dataModel.songCount duration:self.dataModel.duration];
            [headerView addSubview:albumHeader];
            [NSLayoutConstraint activateConstraints:@[
                [albumHeader.leadingAnchor constraintEqualToAnchor:headerView.leadingAnchor],
                [albumHeader.trailingAnchor constraintEqualToAnchor:headerView.trailingAnchor],
                [albumHeader.topAnchor constraintEqualToAnchor:headerView.topAnchor]
            ]];
            
            // Constrain the buttons below the album header
            [playAllAndShuffleHeader.topAnchor constraintEqualToAnchor:albumHeader.bottomAnchor].active = YES;
        } else {
            // Play All and Shuffle buttons only
            [playAllAndShuffleHeader.topAnchor constraintEqualToAnchor:headerView.topAnchor].active = YES;
        }
        
        // Force re-layout using the constraints
        [self.tableView.tableHeaderView layoutIfNeeded];
        self.tableView.tableHeaderView = self.tableView.tableHeaderView;
    }
	
//	self.sectionInfo = self.dataModel.sectionInfo;
//	if (self.sectionInfo)
//		[self.tableView reloadData];
}

#pragma mark Actions

- (IBAction)nowPlayingAction:(id)sender {
    PlayerViewController *playerViewController = [[PlayerViewController alloc] init];
    playerViewController.hidesBottomBarWhenPushed = YES;
	[self.navigationController pushViewController:playerViewController animated:YES];
}

#pragma mark Table view methods

//// Following 2 methods handle the right side index
//- (NSArray *)sectionIndexTitlesForTableView:(UITableView *)tableView  {
//	NSMutableArray *indexes = [[NSMutableArray alloc] init];
//	for (int i = 0; i < self.sectionInfo.count; i++)
//	{
//		[indexes addObject:[[self.sectionInfo objectAtIndexSafe:i] firstObject]];
//	}
//	return indexes;
//}
//
//- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index  {
//	NSUInteger row = [[[self.sectionInfo objectAtIndexSafe:index] objectAtIndexSafe:1] intValue];
//	NSIndexPath *indexPath = [NSIndexPath indexPathForRow:row inSection:0];
//	[tableView scrollToRowAtIndexPath:indexPath atScrollPosition:UITableViewScrollPositionTop animated:NO];
//
//	return -1;
//}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

// Customize the number of rows in the table view.
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section  {
    return section == 0 ? self.dataModel.folderCount : self.dataModel.songCount;
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UniversalTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UniversalTableViewCell.reuseId];
    if (indexPath.section == 0) {
        // Album
        cell.hideSecondaryLabel = YES;
        cell.hideNumberLabel = YES;
        cell.hideCoverArt = NO;
        cell.hideDurationLabel = YES;
        [cell updateWithModel:[self.dataModel folderAlbumWithIndexPath:indexPath]];
    } else {
        // Song
        cell.hideSecondaryLabel = NO;
        cell.hideCoverArt = YES;
        cell.hideDurationLabel = NO;
        ISMSSong *song = [self.dataModel songWithIndexPath:indexPath];
        [cell updateWithModel:song];
        if (song.track == 0) {
            cell.hideNumberLabel = YES;
        } else {
            cell.hideNumberLabel = NO;
            cell.number = song.track;
        }
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
	if (!indexPath) return;
	
    if (indexPath.row < self.dataModel.folderCount) {
        ISMSFolderAlbum *folderAlbum = [self.dataModel folderAlbumWithIndexPath:indexPath];
        FolderAlbumViewController *folderAlbumViewController = [[FolderAlbumViewController alloc] initWithFolderArtist:nil orFolderAlbum:folderAlbum];
        [self pushViewControllerCustom:folderAlbumViewController];
    } else {
        ISMSSong *playedSong = [self.dataModel playSongWithIndexPath:indexPath];
        if (!playedSong.isVideo) {
            [self showPlayer];
        }
    }
}

- (UISwipeActionsConfiguration *)tableView:(UITableView *)tableView trailingSwipeActionsConfigurationForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.dataModel.folderCount) {
        return [SwipeAction downloadAndQueueConfigWithModel:[self.dataModel folderAlbumWithIndexPath:indexPath]];
    } else {
        ISMSSong *song = [self.dataModel songWithIndexPath:indexPath];
        if (!song.isVideo) {
            return [SwipeAction downloadAndQueueConfigWithModel:song];
        }
    }
    return nil;
}

#pragma mark APILoader delegate

- (void)loadingFinished:(APILoader *)loader {
    [viewObjectsS hideLoadingScreen];
    
    [self.tableView reloadData];
    [self addHeaderAndIndex];
    
    [self.refreshControl endRefreshing];
}

- (void)loadingFailed:(APILoader *)loader error:(NSError *)error {
    if (settingsS.isPopupsEnabled) {
        NSString *message = [NSString stringWithFormat:@"There was an error loading the album.\n\nError %li: %@", (long)error.code, error.localizedDescription];
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:message preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
        [self presentViewController:alert animated:YES completion:nil];
    }
	
	[viewObjectsS hideLoadingScreen];
	
    [self.refreshControl endRefreshing];
}

@end

File: Classes/View Controllers/Tabs/Folders/FolderAlbumViewController.swift
//
//  FolderAlbumViewController.swift
//  iSub
//
//  Created by Benjamin Baron on 1/18/21.
//  Copyright © 2021 Ben Baron. All rights reserved.
//

import UIKit
import SnapKit
import Resolver

@objc final class FolderAlbumViewController: UIViewController {
    @Injected private var store: Store
    @Injected private var settings: Settings
    @Injected private var viewObjects: ViewObjects
    
    var serverId = Settings.shared().currentServerId
    
    private let folderArtist: FolderArtist?
    private let folderAlbum: FolderAlbum?
    private let parentFolderId: Int
    private var loader: SubfolderLoader?
    
    private var metadata: FolderMetadata?
    private var folderAlbumIds = [Int]()
    private var songIds = [Int]()
    
    private let tableView = UITableView()
    
    @objc var hasLoaded: Bool { metadata != nil }
    @objc var folderCount: Int { metadata?.folderCount ?? 0 }
    @objc var songCount: Int { metadata?.songCount ?? 0 }
    @objc var duration: Int { metadata?.duration ?? 0 }
    
    @objc init(folderArtist: FolderArtist) {
        self.folderArtist = folderArtist
        self.folderAlbum = nil
        self.parentFolderId = folderArtist.id
        super.init(nibName: nil, bundle: nil)
    }
    
    @objc init(folderAlbum: FolderAlbum) {
        self.folderArtist = nil
        self.folderAlbum = folderAlbum
        self.parentFolderId = folderAlbum.id
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("unimplemented")
    }
    
    deinit {
        cancelLoad()
        NotificationCenter.removeObserverOnMainThread(self)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.backgroundColor = Colors.background
        title = folderArtist?.name ?? folderAlbum?.name
        
        setupDefaultTableView(tableView)
        tableView.refreshControl = RefreshControl { [unowned self] in
            startLoad()
        }
        
        if hasLoaded {
            addHeader()
            addSectionIndex()
        } else {
            startLoad()
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        addShowPlayerButton()
        tableView.reloadData()
        
        NotificationCenter.addObserverOnMainThread(self, selector: #selector(reloadData), name: ISMSNotification_CurrentPlaylistIndexChanged)
        NotificationCenter.addObserverOnMainThread(self, selector: #selector(reloadData), name: ISMSNotification_SongPlaybackStarted)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        cancelLoad()
        
        NotificationCenter.removeObserverOnMainThread(self, name: ISMSNotification_CurrentPlaylistIndexChanged)
        NotificationCenter.removeObserverOnMainThread(self, name: ISMSNotification_SongPlaybackStarted)
    }
    
    // Autolayout solution described here: https://medium.com/@aunnnn/table-header-view-with-autolayout-13de4cfc4343
    private func addHeader() {
        guard folderAlbumIds.count > 0 || songIds.count > 0 else {
            tableView.tableHeaderView = nil
            return
        }
        
        // Create the container view and constrain it to the table
        let headerView = UIView()
        tableView.tableHeaderView = headerView
        headerView.snp.makeConstraints { make in
            make.centerX.width.top.equalToSuperview()
        }
        
        // Create the album header view and constrain to the container view
        let albumHeader: AlbumTableViewHeader?
        if let folderAlbum = folderAlbum {
            albumHeader = AlbumTableViewHeader(folderAlbum: folderAlbum, tracks: songCount, duration: Double(duration))
            headerView.addSubview(albumHeader!)
            albumHeader!.snp.makeConstraints { make in
                make.leading.trailing.top.equalToSuperview()
            }
        } else {
            albumHeader = nil
        }
        
        // Create the play all and shuffle buttons and constrain to the container view
        let playAllAndShuffleHeader = PlayAllAndShuffleHeader(playAllHandler: { [unowned self] in
            SongLoader.playAll(folderId: parentFolderId)
        }, shuffleHandler: { [unowned self] in
            SongLoader.shuffleAll(folderId: parentFolderId)
        })
        headerView.addSubview(playAllAndShuffleHeader)
        playAllAndShuffleHeader.snp.makeConstraints { make in
            if let albumHeader = albumHeader {
                make.top.equalTo(albumHeader.snp.bottom)
            } else {
                make.top.equalToSuperview()
            }
            make.leading.trailing.bottom.equalToSuperview()
        }
        
        // Force re-layout using the constraints
        tableView.tableHeaderView?.layoutIfNeeded()
        tableView.tableHeaderView = tableView.tableHeaderView
    }
    
    private func addSectionIndex() {
        
    }
    
    @objc private func reloadData() {
        tableView.reloadData()
    }
    
    private func loadFromCache() {
        metadata = store.folderMetadata(serverId: serverId, parentFolderId: parentFolderId)
        if metadata != nil {
            folderAlbumIds = store.folderAlbumIds(serverId: serverId, parentFolderId: parentFolderId)
            songIds = store.songIds(serverId: serverId, parentFolderId: parentFolderId)
        } else {
            folderAlbumIds.removeAll()
            songIds.removeAll()
        }
    }
    
    func startLoad() {
        viewObjects.showAlbumLoadingScreen(view, sender: self)
        loader = SubfolderLoader(parentFolderId: parentFolderId) { [weak self] success, error in
            guard let self = self else { return }
            
            if let loader = self.loader {
                self.metadata = loader.folderMetadata
                self.folderAlbumIds = loader.folderAlbumIds
                self.songIds = loader.songIds
            }
            self.loader = nil
            
            if success {
                self.tableView.reloadData()
                self.addHeader()
                self.addSectionIndex()
            } else if let error = error as NSError? {
                if self.settings.isPopupsEnabled {
                    let message = "There was an error loading the album.\n\nError \(error.code): \(error.localizedDescription)"
                    let alert = UIAlertController(title: "Error", message: message, preferredStyle: .alert)
                    alert.addCancelAction(title: "OK")
                    self.present(alert, animated: true, completion: nil)
                }
            }
            self.viewObjects.hideLoadingScreen()
            self.tableView.refreshControl?.endRefreshing()
        }
        loader?.startLoad()
    }
    
    func cancelLoad() {
        loader?.cancelLoad()
        loader?.callback = nil
        loader = nil
        
        tableView.refreshControl?.endRefreshing()
        viewObjects.hideLoadingScreen()
    }
}

extension FolderAlbumViewController: UITableViewConfiguration {
    private func folderAlbum(indexPath: IndexPath) -> FolderAlbum? {
        guard indexPath.row < folderAlbumIds.count else { return nil }
        return store.folderAlbum(serverId: serverId, id: folderAlbumIds[indexPath.row])
    }
    
    private func song(indexPath: IndexPath) -> Song? {
        guard indexPath.row < songIds.count else { return nil }
        return store.song(serverId: serverId, id: songIds[indexPath.row])
    }
    
    private func playSong(indexPath: IndexPath) -> Song? {
        guard indexPath.section == 1, indexPath.row < songIds.count else { return nil }
        return store.playSong(position: indexPath.row, songIds: songIds, serverId: serverId)
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return section == 0 ? folderAlbumIds.count : songIds.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueUniversalCell()
        if indexPath.section == 0 {
            cell.show(cached: false, number: false, art: true, secondary: false, duration: false)
            cell.update(model: folderAlbum(indexPath: indexPath))
        } else {
            if let song = song(indexPath: indexPath) {
                var showNumber = false
                if song.track > 0 {
                    showNumber = true
                    cell.number = song.track
                }
                cell.show(cached: true, number: showNumber, art: false, secondary: true, duration: true)
                cell.update(model: song)
            }
        }
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if indexPath.section == 0 {
            if let folderAlbum = folderAlbum(indexPath: indexPath) {
                let controller = FolderAlbumViewController(folderAlbum: folderAlbum)
                pushViewControllerCustom(controller)
            }
        } else {
            if let song = playSong(indexPath: indexPath), !song.isVideo {
                showPlayer()
            }
        }
    }
    
    func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
        if indexPath.section == 0 {
            return SwipeAction.downloadAndQueueConfig(model: folderAlbum(indexPath: indexPath))
        } else {
            if let song = song(indexPath: indexPath), !song.isVideo {
                return SwipeAction.downloadAndQueueConfig(model: song)
            }
        }
        return nil
    }
}
